Settings Sync: Set Modified Date on settings import

bjtitus opened the pull request:
| **Depends On** | https://github.com/Automattic/pocket-casts-ios/pull/1615 |
|-|-|

Set the Modified Date on all imported settings to 1 second after the epoch date.

* This ensures that imported settings are sent to the server (any settings without a date are not).
* By defaulting to 1 second after, the setting is stored by the server. A default epoch date is effectively a "null" value and will not replace the default value.

## To test

* Change the return value of `shouldEnableSyncedSettings` in `FeatureFlag.swift` to `false`
* Build and run
* Log in to an account which hasn't used synced settings
* Change a few settings
* Enable the `newSettingsStorage` and `settingsSync` feature flags
* Restart the app
* Pull to refresh or tap "Refresh Now" in profile
* Ensure changed settings remain
* Delete the app
* Change `shouldEnableSyncedSettings` in `FeatureFlag.swift` back to its original implementation or set to `true`
* Build and run
* Log in to the same account from above
* Verify that the changed settings are the same as when changed above

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

bjtitus committed:
Add default modifiedAt date when importing

# Conflicts:
#	podcasts/DataManager+Import.swift

bjtitus committed:
Add modifiedAt check to settings import unit tests

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/46">7.63</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

bjtitus committed:
Change default import date to 1 second after epoch

This ensures it doesn't match the server value, which would end up overwriting it.

SergioEstevao commented on podcasts/AppSettings+ImportUserDefaults.swift:
```diff
@@ -6,79 +6,80 @@ extension SettingsStore<AppSettings> {
-        self.update(\.$autoPlayEnabled, value: UserDefaults.standard.bool(forKey: Constants.UserDefaults.autoplay))
-        self.update(\.$notifications, value: UserDefaults.standard.bool(forKey: Constants.UserDefaults.pushEnabled))
-        self.update(\.$appBadge, value: Int32(UserDefaults.standard.integer(forKey: Constants.UserDefaults.appBadge)))
+        let date = Date(timeIntervalSince1970: 1)
```
Maybe `Date.distantPast` will be more clear here.

leandroalonso replied:
@SergioEstevao I don't think we can use that as per:

> By defaulting to 1 second after, the setting is stored by the server. A default epoch date is effectively a "null" value and will not replace the default value.

This would probably result in the setting being discarded by the server.

SergioEstevao replied:
So any reference date will be discarded is this what this means: `A default epoch date` ?

SergioEstevao commented on podcasts/DataManager+Import.swift:
```diff
@@ -5,37 +5,38 @@ extension DataManager {
     func importPodcastSettings() {
         let podcasts = allPodcasts(includeUnsubscribed: true)
 
+        let date = Date(timeIntervalSince1970: 1)
```
Same as above: `Date.distantPast`

SergioEstevao reviewed (approved):
Working good, just a small comment about the way the date is set.

leandroalonso reviewed (changes requested):
@bjtitus this is really not working for me. I'm gonna send you a video in a DM.

bjtitus committed:
Update process settings on full sync

bjtitus committed:
Set podcast syncStatus after importing

MiSikora commented on podcasts/AppSettings+ImportUserDefaults.swift:
```diff
@@ -6,79 +6,80 @@ extension SettingsStore<AppSettings> {
-        self.update(\.$autoPlayEnabled, value: UserDefaults.standard.bool(forKey: Constants.UserDefaults.autoplay))
-        self.update(\.$notifications, value: UserDefaults.standard.bool(forKey: Constants.UserDefaults.pushEnabled))
-        self.update(\.$appBadge, value: Int32(UserDefaults.standard.integer(forKey: Constants.UserDefaults.appBadge)))
+        let date = Date(timeIntervalSince1970: 1)
```
Just double checking if this could be changed to 1 millisecond or should we adjust on Android to 1 second?

bjtitus replied:
I'll get this updated today. I still need to double check the comparisons to make sure the precision is working.

MiSikora commented on podcasts/DataManager+Import.swift:
```diff
@@ -5,40 +5,42 @@ extension DataManager {
     func importPodcastSettings() {
         let podcasts = allPodcasts(includeUnsubscribed: true)
 
+        let date = Date(timeIntervalSince1970: 1)
```
Same as https://github.com/Automattic/pocket-casts-ios/pull/1616#discussion_r1565430956

bjtitus committed:
Sync date extension on Date 1 millisecond after epoch

bjtitus committed:
Adjust unit tests to check for syncDefaultDate

leandroalonso reviewed (changes requested):
@bjtitus the Up Next setting is now persisting correctly, but I still have issues with the Settings under Appearance.

bjtitus committed:
Switch default import date to 1 second after 1970

bjtitus committed:
Use proper default epoch date to match server
